messages/teleporter/message_manager.go

```diff
@@ -105,6 +108,21 @@ func (m *messageManager) ShouldSendMessage(warpMessageInfo *vmtypes.WarpMessageI
 	if !ok {
 		return false, fmt.Errorf("relayer not configured to deliver to destination. destinationChainID=%s", destinationChainID.String())
 	}
+
+	// If allowedDestinations is empty, then all destinations are allowed
+	// If allowedDestinations is not empty, then only the allowed destinations are allowed
+	if len(m.allowedDestinations) > 0 {
+		if allowed, exist := m.allowedDestinations[destinationChainID]; !exist || !allowed {
+			m.logger.Info(
+				"Relayer not configured to relay between source and destination",
+				zap.String("destinationChainID", destinationChainID.String()),
+				zap.String("warpMessageID", warpMessageInfo.WarpUnsignedMessage.ID().String()),
+				zap.String("teleporterMessageID", teleporterMessage.MessageID.String()),
+			)
+			return false, nil
```

Comment:
Yeah, that case should not be an error. With the upcoming subnet-evm changes (as well as anycasting support in the future), I think `ShouldSendMessage` is the right place to check against the list of configured destinations.

Reply:
Thinking about it some more, we can clean up this implementation by using the configured allowed destinations as a filter in on the list of destination clients in `NewRelayer` _before_ the call to `NewMessageManager`. We'd then pass in the filtered list of destination clients. In [relayer.go](https://github.com/ava-labs/awm-relayer/blob/199b3e772ccec3babda3b8cb168109c9afe36f9a/relayer/relayer.go#L78), we'd add something like:

```
var filteredDestinationClients map[ids.ID]ethclient.Client
allowedDestinationChainIDs, err := sourceSubnetInfo.GetAllowedDestinations()
if len(m.allowedDestinations) > 0 { 
    filteredDestinationClients := make(map[ids.ID]ethclient.Client)
    for _, id := range m.allowedDestinations {
        filteredDestinationClients[id] = destinationClients[id]
    }
} else {
    filteredDestinationClients = destinationClients
}
...
messageManager, err := messages.NewMessageManager(logger, addressHash, config, filteredDestinationClients, allowedDestinationChainIDs)
```

This would move logic for deciding which destinations are and are not valid to the relayer constructor which ones once, rather than the message manager `ShouldSendMessage` implementation, which runs for each message. When processing a message log, we could then reference the allowed destinations for that source subnet [here](https://github.com/ava-labs/awm-relayer/blob/199b3e772ccec3babda3b8cb168109c9afe36f9a/relayer/relayer.go#L188)

Reply:
would agree that we should move the destination client filtering logic to `NewRelayer` to prevent checking each message.

Reply:
Moved 👍